### PR TITLE
KAFKA-15853: Move KafkaConfig.Defaults to server module

### PR DESCRIPTION
**Pull request by OmniaGM**
first pr to break #15103 

- [x]  Moving any object companion that KafkaConfigs.Defaults depends on outside of core. 
  - `ClientQuotaManagerConfig`, and `ReplicationQuotaManagerConfig`  object companions moved to `org.apache.kafka.server.config` in another pr need to be merged first 
  - `TransactionLogConfig` object companion moved to `org.apache.kafka.coordinator.transaction`
  - `OffsetConfig` moved to `org.apache.kafka.coordinator.group`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by OmniaGM** [comment]
@mimaison @ijuma can you have a look please?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala [hunk 1]
```diff
@@ -1751,17 +1752,17 @@ object GroupCoordinator {
...
-    loadBufferSize = config.offsetsLoadBufferSize,
-    offsetsRetentionMs = config.offsetsRetentionMinutes * 60L * 1000L,
-    offsetsRetentionCheckIntervalMs = config.offsetsRetentionCheckIntervalMs,
-    offsetsTopicNumPartitions = config.offsetsTopicPartitions,
-    offsetsTopicSegmentBytes = config.offsetsTopicSegmentBytes,
-    offsetsTopicReplicationFactor = config.offsetsTopicReplicationFactor,
-    offsetsTopicCompressionType = config.offsetsTopicCompressionType,
-    offsetCommitTimeoutMs = config.offsetCommitTimeoutMs,
-    offsetCommitRequiredAcks = config.offsetCommitRequiredAcks
+  private[group] def offsetConfig(config: KafkaConfig) = new OffsetConfig(
```
Comment (nizhikov) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: `OffsetConfig` seems not related to `Defaults`.
Shoule we move it to another PR?

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala [hunk 2]
```diff
@@ -51,15 +52,9 @@ case class ClientSensors(metricTags: Map[String, String], quotaSensor: Sensor, t
  *
  */
 case class ClientQuotaManagerConfig(numQuotaSamples: Int =
```
Comment (nizhikov) [inline core/src/main/scala/kafka/server/ClientQuotaManager.scala]: We have new `ClientQuotaManagerConfig.java`.
In the same time scala version of class is here, also.

It seems like we need to keep only one of them.
I think changes related to `ClientQuotaManagerConfig` can be moved to another PR to simplify review and reduce changes.

File: core/src/main/scala/kafka/server/ReplicationQuotaManager.scala [hunk 3]
```diff
@@ -39,18 +39,9 @@ import org.apache.kafka.common.utils.Time
   * @param quotaWindowSizeSeconds     The time span of each sample
   *
   */
-case class ReplicationQuotaManagerConfig(quotaBytesPerSecondDefault: Long = QuotaBytesPerSecondDefault,
```
Comment (nizhikov) [inline core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: We have new `ReplicationQuotaManagerConfig.java`.
In the same time scala version of class is here, also.

It seems like we need to keep only one of them.
I think changes related to `ReplicationQuotaManagerConfig` can be moved to another PR to simplify review and reduce changes.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 1 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Defaults depends on `OffsetConfig` object companion which if we remove it from the `OffsetConfig` file will one have `OffsetConfig` class signature.  I vote to keep it here as it is very small change

Comment (nizhikov) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Thanks. Now I see it.

I prefer to have things as atomic as possible so I prefer to move `OffsetConfig` and other `Defaults` dependencies as separate changes prior to this PR. But it's up to you to decide.

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala (hunk 2 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala]: Defaults depends only on `ClientQuotaManagerConfig` object companion only. Moving `ClientQuotaManagerConfig` class can happened when we move `ClientQuotaManager` or in another pr as this will impact QuotaFactory and some Quota test classes

Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala]: In the JIRA we agreed on only moving the object companions

File: core/src/main/scala/kafka/server/ReplicationQuotaManager.scala (hunk 3 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: Same as above

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala (hunk 2 above)
Comment (nizhikov) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala]: Do we really need instance fields in java version in this case?
Maybe it will be more clear to keep only constants.

File: core/src/main/scala/kafka/server/ReplicationQuotaManager.scala (hunk 3 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: I can move them to another pr before to be merged before this one. If this will make it easier

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala (hunk 2 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala]: > Do we really need instance fields in java version in this case?
Maybe it will be more clear to keep only constants.

remove the contractor make sense for now

File: core/src/main/scala/kafka/server/ReplicationQuotaManager.scala (hunk 3 above)
Comment (nizhikov) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: > If this will make it easier

It will make changes easier in my opinion.
Moving things atomically with clear changes is always better choice.

Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: On a second thought let's break them into another pr because I'll need them for KafkaConfig anyway later

Comment (nizhikov) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: Don't hesitate to ping me for review.

File: core/src/main/scala/kafka/server/ClientQuotaManager.scala (hunk 2 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ClientQuotaManager.scala]: #15159 I moved this to another pr

File: core/src/main/scala/kafka/server/ReplicationQuotaManager.scala (hunk 3 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/server/ReplicationQuotaManager.scala]: done here #15160

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala (hunk 1 above)
Comment (OmniaGM) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: #15161

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.transaction;
+
+public class TransactionLogConfig {
```
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Hey @OmniaGM. It is a bit weird to have those transaction classes in the group-coordinator module. It does not seem to be the correct place.

Comment (OmniaGM) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: I feel the same but also not sure where they should move. They don't fit in server module either. I don't see any Jiras to move transaction coordinator out of server but maybe I can start a new module for transaction coordinator similar to the group one. Would this make more sense?

Comment (ijuma) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Let's see what @dajac says, but I think a `kafka-transactions` or `kafka-transactions-coordinator` module that depends on `kafka-group-coordinator` makes sense to me. The former option is shorter and as clear while the second follows the same pattern as for the group coordinator

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Yeah, creating a new module makes sense to me. I would call it `transaction-coordinator` and the jar would be called `kafka-transaction-coordinator` in order to follow the naming scheme of the `group-coordinator`. I am not sure if it really needs to depend on the group-coordinator module though but I haven't not checked the details.

Comment (ijuma) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Yeah, I was using the maven name. I agree that we shouldn't add a dependency unless it exists.

Comment (OmniaGM) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: I added `transaction-coordinator` modules and moved theses there

**Comment by mimaison** [comment]
@OmniaGM Can we rebase to resolve the conflicts? Thanks

**Comment by OmniaGM** [comment]
> @OmniaGM Can we rebase to resolve the conflicts? Thanks

done now.

**Comment by mimaison** [comment]
@OmniaGM It looks like there are a [few build failures](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15158/6/pipeline/10):
```
[Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15158/core/src/main/scala/kafka/server/DynamicConfig.scala:26:58: Unused import
[Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15158/core/src/main/scala/kafka/server/KafkaConfig.scala:47:50: Unused import
```

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+package org.apache.kafka.coordinator.transaction;
+
+public class TransactionLogConfig {
+    // Log-level config default values
+    public static final int DEFAULT_NUM_PARTITIONS = 50;
+    public static final int DEFAULT_SEGMENT_BYTES = 100 * 1024 * 1024;
+    public static final short DEFAULT_REPLICATION_FACTOR = 3;
+    public static final int DEFAULT_MIN_IN_SYNC_REPLICAS = 2;
+    public static final int DEFAULT_LOAD_BUFFER_SIZE = 5 * 1024 * 1024;
+}
```
Comment (mimaison) [inline transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Nit: We usually have a trailing line at the end of files.

**Comment by OmniaGM** [comment]
> @OmniaGM It looks like there are a [few build failures](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15158/6/pipeline/10):
> 
> ```
> [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15158/core/src/main/scala/kafka/server/DynamicConfig.scala:26:58: Unused import
> [Error] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-15158/core/src/main/scala/kafka/server/KafkaConfig.scala:47:50: Unused import
> ```

fixed this

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java (hunk 5 above)
Comment (OmniaGM) [reply transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java]: Done, and did the same to other files I created as well

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetConfig.java [hunk 6]
```diff
@@ -20,6 +20,7 @@
 import org.apache.kafka.common.record.CompressionType;
 
 public class OffsetConfig {
+
```
Comment (nizhikov) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetConfig.java]: Do we really need this empty line?

File: checkstyle/import-control.xml [hunk 7]
```diff
@@ -261,6 +261,10 @@
         <allow pkg="org.apache.kafka.server.metrics" />
       </subpackage>
     </subpackage>
+
+  <subpackage name="transaction">
```
Comment (nizhikov) [inline checkstyle/import-control.xml]: Do we really need this empty block?

File: transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.transaction;
+
+public class TransactionLogConfig {
```
Comment (nizhikov) [inline transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:19]: Can we introduce new module and config into separate PR?

Comment (OmniaGM) [reply transaction-coordinator/src/main/java/org/apache/kafka/coordinator/transaction/TransactionLogConfig.java:19]: it will be very small one which I don't believe it worth it.

File: checkstyle/import-control-core.xml [hunk 9]
```diff
@@ -82,6 +82,7 @@
     <allow pkg="kafka.log" />
     <allow pkg="kafka.cluster" />
     <allow pkg="kafka.server" />
+    <allow pkg="org.apache.kafka.server" />
```
Comment (mimaison) [inline checkstyle/import-control-core.xml]: Do we really need this change? core seems to build fine without it.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 10]
```diff
@@ -1336,205 +1097,205 @@ object KafkaConfig {
...
+      .define(ReplicationQuotaWindowSizeSecondsProp, INT, Defaults.REPLICATION_QUOTA_WINDOW_SIZE_SECONDS, atLeast(1), LOW, ReplicationQuotaWindowSizeSecondsDoc)
+      .define(AlterLogDirsReplicationQuotaWindowSizeSecondsProp, INT, Defaults.ALTER_LOG_DIRS_REPLICATION_QUOTA_WINDOW_SIZE_SECONDS, atLeast(1), LOW, AlterLogDirsReplicationQuotaWindowSizeSecondsDoc)
+      .define(ControllerQuotaWindowSizeSecondsProp, INT, Defaults.CONTROLLER_QUOTA_WINDOW_SIZE_SECONDS, atLeast(1), LOW, ControllerQuotaWindowSizeSecondsDoc)
       .define(ClientQuotaCallbackClassProp, CLASS, null, LOW, ClientQuotaCallbackClassDoc)
 
       /** ********* General Security Configuration ****************/
-      .define(ConnectionsMaxReauthMsProp, LONG, Defaults.ConnectionsMaxReauthMsDefault, MEDIUM, ConnectionsMaxReauthMsDoc)
-      .define(SaslServerMaxReceiveSizeProp, INT, Defaults.DefaultServerMaxMaxReceiveSize, MEDIUM, SaslServerMaxReceiveSizeDoc)
+      .define(ConnectionsMaxReauthMsProp, LONG, Defaults.CONNECTIONS_MAX_REAUTH_MS, MEDIUM, ConnectionsMaxReauthMsDoc)
+      .define(SaslServerMaxReceiveSizeProp, INT, Defaults.SERVER_MAX_MAX_RECEIVE_SIZE, MEDIUM, SaslServerMaxReceiveSizeDoc)
```
Comment (mimaison) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Can we remove one of the `MAX`?

File: core/src/test/scala/other/kafka/StressTestLog.scala [hunk 11]
```diff
@@ -51,8 +51,8 @@ object StressTestLog {
       scheduler = time.scheduler,
       time = time,
       maxTransactionTimeoutMs = 5 * 60 * 1000,
-      producerStateManagerConfig = new ProducerStateManagerConfig(kafka.server.Defaults.ProducerIdExpirationMs, false),
-      producerIdExpirationCheckIntervalMs = kafka.server.Defaults.ProducerIdExpirationCheckIntervalMs,
+      producerStateManagerConfig = new ProducerStateManagerConfig(org.apache.kafka.server.config.Defaults.PRODUCER_ID_EXPIRATION_MS, false),
```
Comment (mimaison) [inline core/src/test/scala/other/kafka/StressTestLog.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

File: core/src/test/scala/other/kafka/TestLinearWriteSpeed.scala [hunk 12]
```diff
@@ -220,8 +220,8 @@ object TestLinearWriteSpeed {
       brokerTopicStats = new BrokerTopicStats,
       time = Time.SYSTEM,
       maxTransactionTimeoutMs = 5 * 60 * 1000,
-      producerStateManagerConfig = new ProducerStateManagerConfig(kafka.server.Defaults.ProducerIdExpirationMs, false),
-      producerIdExpirationCheckIntervalMs = kafka.server.Defaults.ProducerIdExpirationCheckIntervalMs,
+      producerStateManagerConfig = new ProducerStateManagerConfig(org.apache.kafka.server.config.Defaults.PRODUCER_ID_EXPIRATION_MS, false),
```
Comment (mimaison) [inline core/src/test/scala/other/kafka/TestLinearWriteSpeed.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

File: core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala [hunk 13]
```diff
@@ -270,7 +270,7 @@ class PartitionLockTest extends Logging {
 
     logManager.startup(Set.empty)
     val partition = new Partition(topicPartition,
-      replicaLagTimeMaxMs = kafka.server.Defaults.ReplicaLagTimeMaxMs,
+      replicaLagTimeMaxMs = org.apache.kafka.server.config.Defaults.REPLICA_LAG_TIME_MAX_MS,
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/cluster/PartitionLockTest.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

File: core/src/test/scala/unit/kafka/log/AbstractLogCleanerIntegrationTest.scala [hunk 14]
```diff
@@ -112,8 +112,8 @@ abstract class AbstractLogCleanerIntegrationTest {
         time = time,
         brokerTopicStats = new BrokerTopicStats,
         maxTransactionTimeoutMs = 5 * 60 * 1000,
-        producerStateManagerConfig = new ProducerStateManagerConfig(kafka.server.Defaults.ProducerIdExpirationMs, false),
-        producerIdExpirationCheckIntervalMs = kafka.server.Defaults.ProducerIdExpirationCheckIntervalMs,
+        producerStateManagerConfig = new ProducerStateManagerConfig(org.apache.kafka.server.config.Defaults.PRODUCER_ID_EXPIRATION_MS, false),
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/log/AbstractLogCleanerIntegrationTest.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

File: core/src/test/scala/unit/kafka/log/BrokerCompressionTest.scala [hunk 15]
```diff
@@ -63,8 +63,8 @@ class BrokerCompressionTest {
       time = time,
       brokerTopicStats = new BrokerTopicStats,
       maxTransactionTimeoutMs = 5 * 60 * 1000,
-      producerStateManagerConfig = new ProducerStateManagerConfig(kafka.server.Defaults.ProducerIdExpirationMs, false),
-      producerIdExpirationCheckIntervalMs = kafka.server.Defaults.ProducerIdExpirationCheckIntervalMs,
+      producerStateManagerConfig = new ProducerStateManagerConfig(org.apache.kafka.server.config.Defaults.PRODUCER_ID_EXPIRATION_MS, false),
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/log/BrokerCompressionTest.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

File: core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala [hunk 16]
```diff
@@ -53,7 +53,7 @@ class LogCleanerManagerTest extends Logging {
   val logConfig: LogConfig = new LogConfig(logProps)
   val time = new MockTime(1400000000000L, 1000L)  // Tue May 13 16:53:20 UTC 2014 for `currentTimeMs`
   val offset = 999
-  val producerStateManagerConfig = new ProducerStateManagerConfig(kafka.server.Defaults.ProducerIdExpirationMs, false)
+  val producerStateManagerConfig = new ProducerStateManagerConfig(org.apache.kafka.server.config.Defaults.PRODUCER_ID_EXPIRATION_MS, false)
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala]: Can we import `org.apache.kafka.server.config.Defaults`?

Comment (mimaison) [reply core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala]: I'll stop pointing it for each file but let's import the class.

File: core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala [hunk 17]
```diff
@@ -19,10 +19,9 @@ package kafka.utils
 
 
 import javax.crypto.SecretKeyFactory
-
-import kafka.server.Defaults
 import org.apache.kafka.common.config.ConfigException
 import org.apache.kafka.common.config.types.Password
+import org.apache.kafka.server.config.Defaults
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala:24]: This can be merged with the import 2 lines above

File: server/src/main/java/org/apache/kafka/server/config/Defaults.java [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+
+    /** ********* Group coordinator configuration *********/
+    public static final int GROUP_MIN_SESSION_TIMEOUT_MS = 6000;
+    public static final int GROUP_MAX_SESSION_TIMEOUT_MS = 1800000;
+    public static final int GROUP_INITIAL_REBALANCE_DELAY_MS = 3000;
+    public static final int GROUP_MAX_SIZE = Integer.MAX_VALUE;
+
+    /** ********* New group coordinator configs *********/
+    public static final boolean NEW_GROUP_COORDINATOR_ENABLE = false;
+    public static final List<String> GROUP_COORDINATOR_REBALANCE_PROTOCOLS = Arrays.asList(Group.GroupType.CLASSIC.toString());
```
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/Defaults.java]: Should we use `Collections.singletonList()` here?

File: server/src/main/java/org/apache/kafka/server/config/Defaults.java [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+    public static final short OFFSETS_TOPIC_REPLICATION_FACTOR = OffsetConfig.DEFAULT_OFFSETS_TOPIC_REPLICATION_FACTOR;
+    public static final int OFFSETS_TOPIC_PARTITIONS = OffsetConfig.DEFAULT_OFFSETS_TOPIC_NUM_PARTITIONS;
+    public static final int OFFSETS_TOPIC_SEGMENT_BYTES = OffsetConfig.DEFAULT_OFFSETS_TOPIC_SEGMENT_BYTES;
+    public static final int OFFSETS_TOPIC_COMPRESSION_CODEC = OffsetConfig.DEFAULT_OFFSETS_TOPIC_COMPRESSION_TYPE.id;
+    public static final int OFFSETS_RETENTION_MINUTES = 7 * 24 * 60;
+    public static final long OFFSETS_RETENTION_CHECK_INTERVAL_MS = OffsetConfig.DEFAULT_OFFSETS_RETENTION_CHECK_INTERVAL_MS;
+    public static final int OFFSET_COMMIT_TIMEOUT_MS = OffsetConfig.DEFAULT_OFFSET_COMMIT_TIMEOUT_MS;
+    public static final short OFFSET_COMMIT_REQUIRED_ACKS = OffsetConfig.DEFAULT_OFFSET_COMMIT_REQUIRED_ACKS;
+
+    /** T********* ransaction management configuration *********/
```
Comment (mimaison) [inline server/src/main/java/org/apache/kafka/server/config/Defaults.java]: It looks like something went wrong on this line

Comment (mimaison) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java]: Same on a few other comments below

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. It looks good overall, I just left a few suggestions

File: core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala (hunk 17 above)
Comment (OmniaGM) [reply core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala:24]: The other 2 lines are imports of `org.apache.kafka.common` and not `org.apache.kafka.server`

File: server/src/main/java/org/apache/kafka/server/config/Defaults.java (hunk 19 above)
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java]: fixed

File: server/src/main/java/org/apache/kafka/server/config/Defaults.java (hunk 18 above)
Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java]: yup, updated this

File: core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala (hunk 16 above)
Comment (OmniaGM) [reply core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala]: done

File: checkstyle/import-control-core.xml (hunk 9 above)
Comment (OmniaGM) [reply checkstyle/import-control-core.xml]: nope, I think this was left over from the main pr

File: core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala (hunk 17 above)
Comment (mimaison) [reply core/src/test/scala/unit/kafka/utils/PasswordEncoderTest.scala:24]: Ah yes, my bad!

**Review by mimaison — APPROVED** [review]
LGTM, thanks

File: server/src/main/java/org/apache/kafka/server/config/Defaults.java [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+import org.apache.kafka.server.common.MetadataVersion;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class Defaults {
```
Comment (ijuma) [inline server/src/main/java/org/apache/kafka/server/config/Defaults.java:44]: What's the plan for these defaults, they will not be part of `KafkaConfig` anymore once that's moved to this module?

Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java:44]: 1. They always have been separate object out side of KafkaConfig but in the same file. I just moved them out for reason no. 2
2. Kafkaconfig cant be refactored into Java as one class as it is huge and breaks the complexity rules for the checkstyle. The plan is to keep them out as separate class.

Comment (ijuma) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java:44]: 1. If you look at the pattern in Java, the defaults are usually in the same file as the config file unless it's shared between many configs (then it's extracted).
2. We can add an exception for checkstyle - it's a tool in the end.

Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java:44]: I think the pattern of default and properties in the class object companion has been like this in some scala as well. KafkaConfig has been kinda of anti pattern for a while as the defaults are defined in another object companian. I will raise another pr soon to move this pr to Java pattern.

Comment (OmniaGM) [reply server/src/main/java/org/apache/kafka/server/config/Defaults.java:44]: pr opened here to refactor this #15260
